### PR TITLE
[Feature] Waymo create_data/loading/evaluation speed up

### PR DESCRIPTION
**Pull request by sunjiahao1999**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the back-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdet3d/datasets/convert_utils.py [hunk 1]
```diff
@@ -318,8 +318,12 @@ def get_kitti_style_2d_boxes(info: dict,
 def convert_annos(info: dict, cam_idx: int) -> dict:
     """Convert front-cam anns to i-th camera (KITTI-style info)."""
     rect = info['calib']['R0_rect'].astype(np.float32)
-    lidar2cam0 = info['calib']['Tr_velo_to_cam'].astype(np.float32)
-    lidar2cami = info['calib'][f'Tr_velo_to_cam{cam_idx}'].astype(np.float32)
+    # lidar2cam0 = info['calib']['Tr_velo_to_cam'].astype(np.float32)
```
Comment (Tai-Wang) [inline mmdet3d/datasets/convert_utils.py:321]: Can remove

File: projects/DSVT/configs/dsvt_voxel032_res-second_secfpn_8xb1-cyclic-12e_waymoD5-3d-3class.py [hunk 2]
```diff
@@ -5,7 +5,8 @@
 voxel_size = [0.32, 0.32, 6]
 grid_size = [468, 468, 1]
 point_cloud_range = [-74.88, -74.88, -2, 74.88, 74.88, 4.0]
-data_root = 'data/waymo/kitti_format/'
+data_root = 'data/waymo_mini/kitti_format/'
+# data_root = 'data/waymo/kitti_format/'
```
Comment (Tai-Wang) [inline projects/DSVT/configs/dsvt_voxel032_res-second_secfpn_8xb1-cyclic-12e_waymoD5-3d-3class.py:9]: Should use the 'data/waymo/kitti_format'?

File: projects/DSVT/configs/dsvt_voxel032_res-second_secfpn_8xb1-cyclic-12e_waymoD5-3d-3class.py [hunk 3]
```diff
@@ -212,12 +213,20 @@
 
 val_evaluator = dict(
     type='WaymoMetric',
-    ann_file='./data/waymo/kitti_format/waymo_infos_val.pkl',
-    waymo_bin_file='./data/waymo/waymo_format/gt.bin',
-    data_root='./data/waymo/waymo_format',
+    ann_file='./data/waymo_mini/kitti_format/waymo_infos_val.pkl',
+    waymo_bin_file='./data/waymo_mini/waymo_format/gt_mini.bin',
+    data_root='./data/waymo_mini/waymo_format',
```
Comment (Tai-Wang) [inline projects/DSVT/configs/dsvt_voxel032_res-second_secfpn_8xb1-cyclic-12e_waymoD5-3d-3class.py:218]: use waymo?

File: tools/dataset_converters/kitti_converter.py [hunk 4]
```diff
@@ -109,8 +109,9 @@ def calculate_single(self, info):
         return info
 
     def calculate(self, infos):
-        ret_infos = mmengine.track_parallel_progress(self.calculate_single,
-                                                     infos, self.num_worker)
+        # ret_infos = mmengine.track_parallel_progress(self.calculate_single,
+        #                                              infos, self.num_worker)
```
Comment (Tai-Wang) [inline tools/dataset_converters/kitti_converter.py:113]: remove the unused code

File: tools/dataset_converters/update_infos_to_v2.py [hunk 5]
```diff
@@ -934,24 +934,25 @@ def update_waymo_infos(pkl_path, out_dir):
...
         velo_path = ori_info_dict['point_cloud'].get('velodyne_path')
         if velo_path is not None:
             temp_data_info['lidar_points']['lidar_path'] = Path(velo_path).name
 
         # TODO discuss the usage of Tr_velo_to_cam in lidar
-        Trv2c = ori_info_dict['calib']['Tr_velo_to_cam'].astype(np.float32)
+        # Trv2c = ori_info_dict['calib']['Tr_velo_to_cam'].astype(np.float32)
 
-        temp_data_info['lidar_points']['Tr_velo_to_cam'] = Trv2c.tolist()
+        # temp_data_info['lidar_points']['Tr_velo_to_cam'] = Trv2c.tolist()
```
Comment (Tai-Wang) [inline tools/dataset_converters/update_infos_to_v2.py:946]: remove the unused codes

File: tools/dataset_converters/waymo_converter.py [hunk 6]
```diff
@@ -46,18 +62,11 @@ def __init__(self,
...
-        # Only data collected in specific locations will be converted
-        # If set None, this filter is disabled
-        # Available options: location_sf (main dataset)
-        self.selected_waymo_locations = None
-        self.save_track_id = False
-
+                 save_cam_sync_instances=True,
+                 save_cam_instances=True,
+                 info_prefix='waymo',
+                 max_sweeps=10,
```
Comment (Tai-Wang) [inline tools/dataset_converters/waymo_converter.py:68]: Use 5 to keep it consistent with the original benchmark?

File: tools/dataset_converters/waymo_converter.py [hunk 7]
```diff
@@ -604,29 +400,301 @@ def cart_to_homo(self, mat):
             raise ValueError(mat.shape)
         return ret
 
+    def create_waymo_info_file(self, frame, file_idx, frame_idx, file_infos):
```
Comment (Tai-Wang) [inline tools/dataset_converters/waymo_converter.py:403]: docstring, please also specify the format of info or if we describe the info in this PR message, add a link to this PR

File: tools/dataset_converters/waymo_converter.py [hunk 8]
```diff
@@ -604,29 +400,301 @@ def cart_to_homo(self, mat):
...
+            # Gather cam_instances infos which is used for image-based
+            # (monocular) 3D detection (optional).
+            # TODO: Should we use cam_sync_instances to generate cam_instances?
+            if self.save_cam_instances:
+                frame_infos['cam_instances'] = self.gather_cam_instance_info(
+                    copy.deepcopy(frame_infos['instances']),
+                    frame_infos['images'])
+        file_infos.append(frame_infos)
+
+    def gather_instance_info(self, frame, cam_sync=False):
```
Comment (Tai-Wang) [inline tools/dataset_converters/waymo_converter.py:508]: docstring

Comment (Tai-Wang) [reply tools/dataset_converters/waymo_converter.py:508]: the input arg cam_sync should be replaced with self.xxxx?

File: tools/dataset_converters/waymo_converter.py [hunk 9]
```diff
@@ -604,29 +400,301 @@ def cart_to_homo(self, mat):
...
+                 rotation_y]).astype(np.float32).tolist()
+            instance_info['bbox_label_3d'] = label
+            instance_info['num_lidar_pts'] = obj.num_lidar_points_in_box
+
+            if self.save_track_id:
+                instance_info['track_id'] = obj.id
+            instance_infos.append(instance_info)
+        return instance_infos
+
+    def gather_cam_instance_info(self, instances: dict, images: dict):
```
Comment (Tai-Wang) [inline tools/dataset_converters/waymo_converter.py:593]: docstring

File: tools/dataset_converters/waymo_converter.py [hunk 10]
```diff
@@ -604,29 +400,301 @@ def cart_to_homo(self, mat):
...
+
 
 def create_ImageSets_img_ids(root_dir, splits):
+    """Create txt files indicating what to collect in each split."""
     save_dir = join(root_dir, 'ImageSets/')
     if not exists(save_dir):
         os.mkdir(save_dir)
 
-    idx_all = [[] for i in splits]
+    idx_all = [[] for _ in range(4)]
```
Comment (Tai-Wang) [inline tools/dataset_converters/waymo_converter.py:681]: 4 is hard code

File: tools/create_data.py [hunk 11]
```diff
@@ -171,23 +171,40 @@ def waymo_data_prep(root_path,
                     version,
                     out_dir,
                     workers,
-                    max_sweeps=5):
-    """Prepare the info file for waymo dataset.
+                    max_sweeps=10,
```
[the remark this answers is not in the document]
Comment (sunjiahao1999) [reply tools/create_data.py:174]: We use 1/5 infos when load_data_list in WaymoDataset https://github.com/open-mmlab/mmdetection3d/blob/0f9dfa97a35ef87e16b700742d3c358d0ad15452/mmdet3d/datasets/waymo_dataset.py#L187

File: tools/dataset_converters/waymo_converter.py (hunk 6 above)
Comment (sunjiahao1999) [reply tools/dataset_converters/waymo_converter.py:71]: It's the max sweeps. We can only use 5 sweeps when we load sweeps information.

File: tools/dataset_converters/waymo_converter.py (hunk 8 above)
Comment (sunjiahao1999) [reply tools/dataset_converters/waymo_converter.py:519]: We will use both `cam_sync=False` and `cam_sync=True` for generating infos of ['instances'] and ['cam_sync_instances']

**Pull request closed without merging**